src/twisted/python/deprecate.py

```diff
@@ -714,10 +787,15 @@ def function(tweedledum=1, tweedledee=2):
     @rtype: 1-argument callable taking a callable and returning a callable.
     """
     def wrapper(wrappee):
-        argspec = inspect.getargspec(wrappee)
+        if hasattr(inspect, "getfullargspec"):
```

Comment:
As a first approximation, don't ever use `hasattr`.  It masks exceptions which can produce confusing behavior.

Generally `getattr(inspect, name, None)` is a good replacement.

In this case, since this pattern is repeated in the test module, perhaps you want to stick something into `twisted.python.compat` for "non-deprecated get-a-signature-like-function" and use that directly from each of these callers.